packages/ui/react-ui-data/src/hooks/useForm.ts

```diff
@@ -4,54 +4,87 @@
 
 import { type ChangeEvent, type FocusEvent, useCallback, useState } from 'react';
 
-import { type ValidationError } from '../util';
+import { type S } from '@dxos/effect';
+import { invariant } from '@dxos/invariant';
 
-// TODO(Zan): This module will be generally useful and we should factor it out after the
-// API crystallizes a bit more.
+import { validateSchema, type ValidationError } from '../util';
+
+// TODO(ZaymonFC): Instead of delegating validation of a schema to the validate call
+// you should be able to provide a schema for `values`. Note that there should still
+// be a facility for custom validation logic.
+
+type InputValue = string | number | readonly string[] | undefined;
 
 interface FormOptions<T> {
   initialValues: T;
-  validate: (values: any) => ValidationError[] | undefined;
+  schema?: S.Schema<T>;
+  /**
+   * Custom validation function that runs only after schema validation passes.
+   * Use this for complex validation logic that can't be expressed in the schema.
+   * @returns Array of validation errors, or undefined if validation passes
+   */
+  additionalValidation?: (values: T) => ValidationError[] | undefined;
```

Comment:
```suggestion
  validate?: (values: T) => ValidationError[] | undefined;
```

Reply:
We use the provided schema for the first pass of validation, and if that passes then this function has a chance to do additional refinements so we need a name to reflect that.